### PR TITLE
New CS: Database Security

### PR DESCRIPTION
**Pull request by rbsec**
Draft PR based to create a new cheat sheet for database security, as per #293.

This PR is to track progress and allow discussion. Any suggestions on content, things to add, etc are very welcome.

It is still in the draft stages, so please hold doing a syntactic review until the initial draft is complete.

- [x] Introduction
- [x] Contents
- [x] Connecting to the database
  - [x] Transport layer protection
- [x] Authentication
- [x] Permissions
  - [x] Advanced Permissions
- [x] Database configuration and hardening
- [x] Database specific recommendations
  - [x] MySQL
  - [x] Microsoft SQL Server

### CODE REVIEW

**Comment by rbsec** [comment]
Initial draft of the CS is done - would be great to have a content-based review to see if there's anything that you all disagree with, or that you think should be added.

The DB specific stuff is a bit sparse - I've mostly used MySQL/SQL Server in the base, so happy for some additions for other DBs. Should be key things that aren't covered elsewhere in the CS.

**Comment by ThunderSon** [comment]
Please don't remove the name of the CS from the top. It is better for PDF publishing once we get to that stage, and on the website, it will show up on the top, which is better 😄

**Comment by rbsec** [comment]
@mackowski I've updated the TLS guidance

File: cheatsheets/Database_Security_Cheat_Sheet.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
+# Database Security Cheat Sheet
+
+## Introduction
+
+This cheat sheet provides guidance on securely configuring and using SQL databases. It is intended to be used by application developers when they are responsible for managing the databases, in the absence of a dedicated database administrator (DBA). For details about protecting against SQL Injection attacks, see the [SQL Injection Prevention Cheat Sheet](SQL_Injection_Prevention_Cheat_Sheet.md).
+
+## Contents
+
+- [Introduction](#introduction)
```
[the remark this answers is not in the document]
Comment (rbsec) [reply cheatsheets/Database_Security_Cheat_Sheet.md]: Removed.

File: cheatsheets/Database_Security_Cheat_Sheet.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+The backend database used by the application should be isolated as much as possible, in order to prevent malicious or undesirable users from being able to connect to it. Exactly how this is achieved will depend on the system and network architecture. The following options could be used to protect it:
+
+- Disabling network (TCP) access and requiring all access is over a local socket file or named pipe.
+- Configuring the database to only bind on localhost.
+- Restricting access to the network port to specific hosts with firewall rules.
+- Placing the database server in a separate DMZ isolated from the application server.
+
+Similar protection should be implemented to protect any web-based management tools used with the database, such as phpMyAdmin.
+
+Connections should never be allowed to the database from applications running on untrusted systems (such as a thick-client application that connects directly to the database, rather than using an API).
```
[the remark this answers is not in the document]
Comment (rbsec) [reply cheatsheets/Database_Security_Cheat_Sheet.md]: Updated in db467ea - does the new wording look ok?

File: cheatsheets/Database_Security_Cheat_Sheet.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+- Do not use the built in `root` or `sa` or `SYS` accounts.
+- Do not grant the account administrative rights over the database instance.
+- Only allow the account to connect from whitelisted hosts.
+  - This would often be `localhost` or the address of the application server.
+- Only grant the account access to the specific databases it needs.
+  - Development, UAT and Production environments should all use separate databases and accounts.
+- Only grant the required permissions on the databases.
+  - Most applications would only need `SELECT`, `UPDATE` and `DELETE` permissions.
+  - The account should not be the owner of the database as this lead to privilege escalation vulnerabilities.
```
Comment (ThunderSon) [inline cheatsheets/Database_Security_Cheat_Sheet.md]: leads instead of lead

File: cheatsheets/Database_Security_Cheat_Sheet.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+The backend database used by the application should be isolated as much as possible, in order to prevent malicious or undesirable users from being able to connect to it. Exactly how this is achieved will depend on the system and network architecture. The following options could be used to protect it:
+
+- Disabling network (TCP) access and requiring all access is over a local socket file or named pipe.
+- Configuring the database to only bind on localhost.
+- Restricting access to the network port to specific hosts with firewall rules.
+- Placing the database server in a separate DMZ isolated from the application server.
+
+Similar protection should be implemented to protect any web-based management tools used with the database, such as phpMyAdmin.
+
+When an application is running on an untrusted system (such as a thick-client), it should always connect to the backend through an API that can enforce appropriate access control and restrictions. Direct connections should **never** be made from a thick client to the backend database.
```
Comment (ThunderSon) [inline cheatsheets/Database_Security_Cheat_Sheet.md]: access controls, plural

Comment (rbsec) [reply cheatsheets/Database_Security_Cheat_Sheet.md]: "Access control" is valid, it's referring to the concept, rather than individual controls.

**Pull request merged**